### ISSUE TITLE
why use GT

### ISSUE THREAD

**Issue by Tomwmg** [cocodataset/cocoapi#337]

**Comment by Tomwmg** [comment]
First, I acoording the GT to generate the test format results, and I set the score=1.0.
When I use this api to test val2014 and val2017 in keypoints task with gt-result , following results are generated. I don't know what's right. Can someone explain it to me, thank you.
2014val
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.585
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.609
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.578
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
2017val
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.751
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.751
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.751
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000

**Comment by niazahamd89** [comment]
Hi @Tomwmg i need a little information how you generate the Dt result. json file for evaluation let me explain that please.

**Comment by niazahamd89** [comment]
Hi @Tomwmg i need a little information how you generate Dt result.json file let me explain that please.

I want to clear is i train my model on coco train and validation set for pose estimation now i want to evaluate my model to find the AP and AR. i am going to use coco evaluation metric for this purpose, in demo.ipynb the first file i need is the Gt file which i downloaded from coco site the second file **i need the Dt file now i am confuse how i can generate this Dt file**? my training is complete the model generate weights file in .ckpt format and i test the model is able to plot pose in pictures.

there are some result.json files in cocoapi result folder https://github.com/philferriere/cocoapi/tree/master/results basically the same file person_keypoints_val2014_fakekeypoints100_results.json i want to generate for my model to evaluate.

**Comment by ZP-Guo** [comment]
Hi, @niazahamd89, 
I have the same question as @Tomwmg. 
I want to generate keypoints annotations json file and evaluate it. And then I use GT file to generate result file to test the cocoapi. My json likes `{"category_id": 1, "image_id": 425226, "score": 0, "keypoints": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 142, 309, 1, 177, 320, 2, 191, 398, 2, 237, 317, 2, 233, 426, 2, 306, 233, 2, 92, 452, 2, 123, 468, 2, 0, 0, 0, 251, 469, 2, 0, 0, 0, 162, 551, 2]}`. My result is 
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.641
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.643
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.641
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000.
I use the GT "keypoints" and set "score" is "0". I do not what function "score" is. Could you tell.
I am looking forward your reply. Thanks.

**Comment by niazahamd89** [comment]
Hi @ZP-Guo would you plz clear me how you generate your pose result in this format   `{"category_id": 1, "image_id": 425226, "score": 0, "keypoints": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 142, 309, 1, 177, 320, 2, 191, 398, 2, 237, 317, 2, 233, 426, 2, 306, 233, 2, 92, 452, 2, 123, 468, 2, 0, 0, 0, 251, 469, 2, 0, 0, 0, 162, 551, 2]}`
is there any proper standard code that generates this array.

**Comment by niazahamd89** [comment]
i got you! you manually by hand type it from GT file. i am also confuse how i can get this result after plotting pose on picture and save this whole information. i think that would be your result but how to do it i am also digging the solution from 3 weeks. there should be a proper code that feed your model plus evl images and generate this reul.json file at output do you have any idea?

**Comment by ZP-Guo** [comment]
> i got you! you manually by hand type it from GT file. i am also confuse how i can get this result after plotting pose on picture and save this whole information. i think that would be your result but how to do it i am also digging the solution from 3 weeks. there should be a proper code that feed your model plus evl images and generate this reul.json file at output do you have any idea?

According to this json file [https://github.com/cocodataset/cocoapi/blob/master/results/person_keypoints_val2014_fakekeypoints100_results.json](url), I generate result.json by hand and the format of my json file is as same as its.     So you need to know where you can get the prediction pose coordinates in your code if you want to do it as I do. 
I do not go to find the code you call proper codes, because I think I can do it by myself. Maybe you too.

**Comment by niazahamd89** [comment]
i test a sample example of result file as you did i change the score 0.0, 0.29, 0.1,1.0 but changing do't  effect on AP or AR may be you tried it. But i think the result file which is generated from pose estimation model under proper standard method which is responsible to generated these detected pose results, will also generate the `score` value. 

Basically my problem is i want to generated that `keypoints` coordinates which my model plot on picture. i am confuse how to generate these, not to pick from GT file by hand. 

I have a question here is this the proper way for evaluation to make this result.json file by hand? do you have any idea?

**Comment by edspiration** [comment]
Hi I also observe similar problems as @Tomwmg and @ZP-Guo 

I tried to validate ground truth against itself.

The code look something like this:

```
PERSON_CAT_ID = 1
person_ann_ids = cocoGt.getAnnIds(catIds=[PERSON_CAT_ID])
person_anns = cocoGt.loadAnns(ids=person_ann_ids)

results = []

for person_ann in person_anns:
    results.append({
        "image_id": person_ann['image_id'],
        "category_id": PERSON_CAT_ID,
        'keypoints': person_ann['keypoints'],
        'bbox': person_ann['bbox'],
        'score': 1.0
    })
    
with open("results.txt", "w") as f:
    f.write(json.dumps(results))
    
cocoDt = cocoGt.loadRes("results.txt")
cocoEval = COCOeval(cocoGt, cocoDt, 'keypoints')
cocoEval.evaluate()
cocoEval.accumulate()
cocoEval.summarize()
```

This is the results I get when I expected 100% for all AP:

```
Loading and preparing results...
DONE (t=0.06s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=4.89s).
Accumulating evaluation results...
DONE (t=0.08s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.641
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.643
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.641
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.814
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.963
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
```

I tried 2 other combinations that I can't understand the returned results:

1. I only include annotated ground truth when I save to results file. 

```
for person_ann in person_anns:
    if person_ann['num_keypoints'] > 0:
        results.append({
            "image_id": person_ann['image_id'],
            "category_id": PERSON_CAT_ID,
            'keypoints': person_ann['keypoints'],
            'bbox': person_ann['bbox'],
            'score': 1.0
        })
```

The results i get has 0.999 for **Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ]**:

```
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.999
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
```


2. I only include annotated ground truth when I save to results file AND I don't include 'bbox'
```
for person_ann in person_anns:
    if person_ann['num_keypoints'] > 0:
        results.append({
            "image_id": person_ann['image_id'],
            "category_id": PERSON_CAT_ID,
            'keypoints': person_ann['keypoints'],
            'score': 1.0
        })
```
I was finally getting 100% for all APs and ARs:

```
Loading and preparing results...
DONE (t=0.16s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=2.44s).
Accumulating evaluation results...
DONE (t=0.04s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 1.000
```

From the coco dataset [site](http://cocodataset.org/#keypoints-eval), I can't understand why 'bbox' can affect the AP and AR scores.

**Comment by sandhyacs** [comment]
Hi,

I want to know, what is this GT file?
